File: python/greedy_algorithms.py
fruits = set(['avocado', 'tomato', 'banana'])
vegetables = set(['beets', 'carrots', 'tomato'])

# Union
print(fruits | vegetables)

# Intersection
print(fruits & vegetables)

# Difference
print(fruits - vegetables)
print(vegetables - fruits)

# Example
states_needed = set(["a", "b", "c", "d", "e", "f", "g", "h"])

stations = {}
stations["kone"] = set(["d", "e", "f"])
stations["ktwo"] = set(["b", "d", "a"])
stations["kthree"] = set(["c", "e", "g"])
stations["kfour"] = set(["e", "f"])
stations["kfive"] = set(["g", "h"])

final_stations = set()

while states_needed:
    best_station = None
    states_covered = set()
    for station, states_for_station in stations.items():
        covered = states_needed & states_for_station
        if len(covered) > len(states_covered):
            best_station = station
            states_covered = covered

    states_needed -= states_covered
    final_stations.add(best_station)

print(final_stations)

File: javascript/greedyAlgorithms.js
let fruits = new Set(["avocado", "tomato", "banana"]);
let vegetables = new Set(["beets", "carrots", "tomato"]);

// Union
const union = (setA, setB) => {
  let setC = new Set();

  for (let element of setA) {
    setC.add(element);
  }
  for (let element of setB) {
    setC.add(element);
  }
  return setC;
};

console.log(union(fruits, vegetables));

// Intersection
const intersection = (setA, setB) => {
  let setC = new Set();

  for (let element of setB) {
    if (setA.has(element)) {
      setC.add(element);
    }
  }
  return setC;
};

console.log(intersection(fruits, vegetables));

// Difference
const difference = (setA, setB) => {
  let setC = new Set(setA);

  for (let element of setB) {
    setC.delete(element);
  }
  return setC;
};

console.log(difference(fruits, vegetables));
console.log(difference(vegetables, fruits));

// Example
let statesNeeded = new Set(["a", "b", "c", "d", "e", "f", "g", "h"]);

const stations = {};
stations.kone = new Set(["d", "e", "f"]);
stations.ktwo = new Set(["b", "d", "a"]);
stations.kthree = new Set(["c", "e", "g"]);
stations.kfour = new Set(["e", "f"]);
stations.kfive = new Set(["g", "h"]);

const finalStations = new Set();

while (statesNeeded.size) {
  let bestStation = null;
  let statesCovered = new Set();
  Object.keys(stations).forEach((station) => {
    const states = stations[station];
    const covered = new Set([...statesNeeded].filter((x) => states.has(x)));
    if (covered.size > statesCovered.size) {
      bestStation = station;
      statesCovered = covered;
    }
  });
  statesNeeded = new Set(
    [...statesNeeded].filter((x) => !statesCovered.has(x))
  );
  finalStations.add(bestStation);
}

console.log(finalStations);
